File: scripts/tasks/sync-remotes.ts
import { readdirSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, join } from "node:path";
import { isGitRepo, logStep, runOrThrow } from "../common";

const DEFAULTS = {
  root: join(homedir(), "github"),
  owner: "dunamismax",
  ghAlias: "github.com-dunamismax",
  cbAlias: "codeberg.org-dunamismax",
};

type RepoResult = {
  name: string;
  status: "ok" | "skip" | "fixed" | "error";
  detail?: string;
};

function getRemoteUrls(
  repoPath: string,
): { fetchUrl: string; pushUrls: string[] } | null {
  try {
    const fetchUrl = runOrThrow(
      ["git", "-C", repoPath, "remote", "get-url", "origin"],
      { quiet: true },
    );
    const pushRaw = runOrThrow(
      ["git", "-C", repoPath, "remote", "get-url", "--push", "--all", "origin"],
      { quiet: true },
    );
    const pushUrls = pushRaw
      .split("\n")
      .map((l) => l.trim())
      .filter(Boolean);
    return { fetchUrl, pushUrls };
  } catch {
    return null;
  }
}

function configureRemote(repoPath: string, ghUrl: string, cbUrl: string): void {
  runOrThrow(["git", "-C", repoPath, "remote", "set-url", "origin", ghUrl], {
    quiet: true,
  });

  // Clear all push URLs then re-add both
  try {
    runOrThrow(
      ["git", "-C", repoPath, "config", "--unset-all", "remote.origin.pushurl"],
      { quiet: true },
    );
  } catch {
    // No pushurl config to clear — that's fine
  }

  runOrThrow(
    [
      "git",
      "-C",
      repoPath,
      "remote",
      "set-url",
      "--add",
      "--push",
      "origin",
      ghUrl,
    ],
    { quiet: true },
  );
  runOrThrow(
    [
      "git",
      "-C",
      repoPath,
      "remote",
      "set-url",
      "--add",
      "--push",
      "origin",
      cbUrl,
    ],
    { quiet: true },
  );
}

function expectedUrls(repoName: string): { ghUrl: string; cbUrl: string } {
  return {
    ghUrl: `git@${DEFAULTS.ghAlias}:${DEFAULTS.owner}/${repoName}.git`,
    cbUrl: `git@${DEFAULTS.cbAlias}:${DEFAULTS.owner}/${repoName}.git`,
  };
}

export function isCorrect(
  urls: { fetchUrl: string; pushUrls: string[] },
  expected: { ghUrl: string; cbUrl: string },
): boolean {
  if (urls.fetchUrl !== expected.ghUrl) return false;
  if (urls.pushUrls.length !== 2) return false;
  if (urls.pushUrls[0] !== expected.ghUrl) return false;
  if (urls.pushUrls[1] !== expected.cbUrl) return false;
  return true;
}

function processRepo(repoPath: string, fix: boolean): RepoResult {
  const name = basename(repoPath);

  if (!isGitRepo(repoPath)) {
    return { name, status: "skip", detail: "not a git repo" };
  }

  const urls = getRemoteUrls(repoPath);
  if (!urls) {
    return { name, status: "skip", detail: "no origin remote" };
  }

  const expected = expectedUrls(name);

  if (isCorrect(urls, expected)) {
    return { name, status: "ok" };
  }

  if (!fix) {
    const issues: string[] = [];
    if (urls.fetchUrl !== expected.ghUrl) {
      issues.push(`fetch: ${urls.fetchUrl} (want ${expected.ghUrl})`);
    }
    if (urls.pushUrls.length !== 2) {
      issues.push(`push url count: ${urls.pushUrls.length} (want 2)`);
    } else {
      if (urls.pushUrls[0] !== expected.ghUrl) {
        issues.push(`push[0]: ${urls.pushUrls[0]} (want ${expected.ghUrl})`);
      }
      if (urls.pushUrls[1] !== expected.cbUrl) {
        issues.push(`push[1]: ${urls.pushUrls[1]} (want ${expected.cbUrl})`);
      }
    }
    return { name, status: "error", detail: issues.join("; ") };
  }

  try {
    configureRemote(repoPath, expected.ghUrl, expected.cbUrl);
    return { name, status: "fixed" };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return { name, status: "error", detail: msg };
  }
}

function discoverRepos(root: string): string[] {
  return readdirSync(root)
    .map((entry) => join(root, entry))
    .filter((p) => {
      try {
        return statSync(p).isDirectory();
      } catch {
        return false;
      }
    })
    .sort();
}

export function syncRemotes(): void {
  const fix = Bun.argv.includes("--fix");
  const root = DEFAULTS.root;

  logStep(
    fix
      ? "Syncing dual remotes (fix mode)"
      : "Checking dual remotes (dry run — pass --fix to apply)",
  );

  console.log(`  root:     ${root}`);
  console.log(`  owner:    ${DEFAULTS.owner}`);
  console.log(`  github:   ${DEFAULTS.ghAlias}`);
  console.log(`  codeberg: ${DEFAULTS.cbAlias}`);

  const repos = discoverRepos(root);
  const results = repos.map((r) => processRepo(r, fix));

  console.log("");

  const ok = results.filter((r) => r.status === "ok");
  const fixed = results.filter((r) => r.status === "fixed");
  const skipped = results.filter((r) => r.status === "skip");
  const errors = results.filter((r) => r.status === "error");

  for (const r of ok) console.log(`  [ok]    ${r.name}`);
  for (const r of fixed) console.log(`  [fixed] ${r.name}`);
  for (const r of skipped) console.log(`  [skip]  ${r.name} — ${r.detail}`);
  for (const r of errors) console.log(`  [error] ${r.name} — ${r.detail}`);

  console.log(
    `\n  ok=${ok.length} fixed=${fixed.length} skipped=${skipped.length} errors=${errors.length}`,
  );

  if (errors.length > 0) {
    process.exit(1);
  }
}

File: scripts/tasks/sync_remotes.py
"""Safely inspect and normalize git remote policies under ~/github."""

from __future__ import annotations

import sys
from dataclasses import dataclass
from pathlib import Path

from scripts.common import is_git_repo, log_step
from scripts.remote_policy import (
    apply_repo_remote_policy,
    build_repo_remote_policy,
    get_remote_push_default,
    get_remote_urls,
    list_git_remotes,
    repo_url_matches,
)

DEFAULTS = {
    "root": str(Path.home() / "github"),
    "owner": "dunamismax",
    "gh_alias": "github.com-dunamismax",
    "cb_alias": "codeberg.org-dunamismax",
    "openclaw_upstream_url": "https://github.com/openclaw/openclaw.git",
}


@dataclass
class RepoResult:
    name: str
    status: str  # "ok" | "todo" | "skip" | "fixed" | "error"
    detail: str = ""


def _is_correct(fetch_url: str, push_urls: list[str], gh_url: str, cb_url: str) -> bool:
    if fetch_url != gh_url:
        return False
    if len(push_urls) != 2:
        return False
    return push_urls[0] == gh_url and push_urls[1] == cb_url


def _custom_topology_reason(repo_path: str, repo_name: str, fetch_url: str) -> str | None:
    remotes = sorted(list_git_remotes(repo_path))
    push_default = get_remote_push_default(repo_path)

    if push_default and push_default != "origin":
        return f"custom pushDefault={push_default}"

    extra_remotes = [remote for remote in remotes if remote != "origin"]
    if extra_remotes:
        return f"custom remotes: {', '.join(extra_remotes)}"

    if not repo_url_matches(fetch_url, DEFAULTS["owner"], repo_name):
        return f"origin fetch preserved: {fetch_url}"

    return None


def _process_repo(repo_path: str, fix: bool) -> RepoResult:
    name = Path(repo_path).name

    if not is_git_repo(repo_path):
        return RepoResult(name, "skip", "not a git repo")

    urls = get_remote_urls(repo_path)
    if urls is None:
        return RepoResult(name, "skip", "no origin remote")

    fetch_url, push_urls = urls
    policy = build_repo_remote_policy(
        name,
        owner=DEFAULTS["owner"],
        github_host_alias=DEFAULTS["gh_alias"],
        codeberg_host_alias=DEFAULTS["cb_alias"],
        openclaw_upstream_url=DEFAULTS["openclaw_upstream_url"],
    )
    custom_reason = _custom_topology_reason(repo_path, name, fetch_url)

    if len(policy.origin.push_urls) == 2:
        gh_url, cb_url = policy.origin.push_urls
        if _is_correct(fetch_url, push_urls, policy.origin.fetch_url, cb_url):
            return RepoResult(name, "ok")

        if custom_reason is not None:
            return RepoResult(name, "skip", custom_reason)

        if not fix:
            issues: list[str] = []
            if fetch_url != policy.origin.fetch_url:
                issues.append(f"fetch: {fetch_url} (want {policy.origin.fetch_url})")
            if len(push_urls) != 2:
                issues.append(f"push url count: {len(push_urls)} (want 2)")
            else:
                if push_urls[0] != gh_url:
                    issues.append(f"push[0]: {push_urls[0]} (want {gh_url})")
                if push_urls[1] != cb_url:
                    issues.append(f"push[1]: {push_urls[1]} (want {cb_url})")
            return RepoResult(name, "todo", "; ".join(issues))

        try:
            apply_repo_remote_policy(repo_path, policy)
            return RepoResult(name, "fixed")
        except RuntimeError as exc:
            return RepoResult(name, "error", str(exc))

    return RepoResult(name, "skip", custom_reason or "non-mirror policy repo")


def _discover_repos(root: str) -> list[str]:
    root_path = Path(root)
    return sorted(
        str(path)
        for path in root_path.iterdir()
        if path.is_dir() and not path.name.startswith(".")
    )


def sync_remotes() -> None:
    fix = "--fix" in sys.argv
    root = DEFAULTS["root"]

    log_step(
        "Syncing managed mirror remotes (fix mode)"
        if fix
        else "Checking managed mirror remotes (dry run — pass --fix to apply)"
    )

    print(f"  root:     {root}")
    print(f"  owner:    {DEFAULTS['owner']}")
    print(f"  github:   {DEFAULTS['gh_alias']}")
    print(f"  codeberg: {DEFAULTS['cb_alias']}")

    repos = _discover_repos(root)
    results = [_process_repo(repo, fix) for repo in repos]

    print()

    ok = [result for result in results if result.status == "ok"]
    todo = [result for result in results if result.status == "todo"]
    fixed = [result for result in results if result.status == "fixed"]
    skipped = [result for result in results if result.status == "skip"]
    errors = [result for result in results if result.status == "error"]

    for result in ok:
        print(f"  [ok]    {result.name}")
    for result in todo:
        print(f"  [todo]  {result.name} — {result.detail}")
    for result in fixed:
        print(f"  [fixed] {result.name}")
    for result in skipped:
        print(f"  [skip]  {result.name} — {result.detail}")
    for result in errors:
        print(f"  [error] {result.name} — {result.detail}")

    print(
        f"\n  ok={len(ok)} todo={len(todo)} fixed={len(fixed)} skipped={len(skipped)} errors={len(errors)}"
    )

    if errors:
        sys.exit(1)
